### PR TITLE
Add OpenSergoClientManager to support client reuse

### LINKED ISSUE #8: [DISCUSSION] Best practice of shared/individual client between different integrations in the same process | 同个应用进程多个框架对接 OpenSergo 时 client 的最佳实践
我们的微服务通常会使用好几种框架，比如 RPC 使用的 Spring Cloud Alibaba，流量治理使用的 Sentinel，数据库中间件使用的 ShardingSphere。在这些框架都对接了 OpenSergo 的情况下，在接入到 OpenSergo 控制面，对 OpenSergo client 的使用会有几种方式：

* 每个框架对接 OpenSergo SDK 的时候，分别创建各自的 client 来接入控制面，各自监听自己需要的配置。这样做，不同传输链路不会相互影响；弊端就是同个进程多条长连接，不利于控制面连接管理，以及整体的性能。
* 不同框架对接 OpenSergo 时 share 同个 OpenSergo client，也即 share 同个长连接（正常情况下连接的控制面都是一个）。不同框架只需要往同个 client 注册不同 subscriber 即可。这样的好处是只需要一个长连接，可以比较好地进行复用。

目前的实现未预留 shared 的设计，社区可以一起讨论下这几种情况的最佳实践，以及 shared client 的设计。

### PR DESCRIPTION
**Pull request by jnan806**
Signed-off-by: Jiangnan Jia <jnan0806@gmail.com>

### Describe what this PR does / why we need it
support client reuse

### Does this pull request fix one issue?

Fixes #8 

### Describe how you did it
- Add `OpenSergoClientManager` to manage the reuse OpenSergoClient
- Add `Builder` in OpenSergoClient for creating instance
- Add `OpenSergoConfig` to support customized config

### Describe how to verify it
- for shared OpenSergoClient
``` java
OpenSergoClient openSergoClient = OpenSergoClientManager.get().getOrCreateClient("127.0.0.1", 10246, new OpenSergoConfig());
```
- for prototype OpenSergoClient
``` java 
OpenSergoClient openSergoClient = new OpenSergoClient.Builder().endpoint("127.0.0.1", 
10246).openSergoConfig(new OpenSergoConfig()).build();
```

### Special notes for reviews

### CODE REVIEW

File: src/main/java/io/opensergo/OpenSergoClientManager.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    /**
+     * get the instance from sharedOpenSergoClientCache with config.
+     * If instance can be found by host and port, the one will be returned, whether the config is matched or not.
+     *
+     * @param host endpoint of the OpenSergo Control Plane
+     * @param port port of the OpenSergo Control Plane
+     * @param config OpenSergoConfig
+     * @return OpenSergoClient
+     */
+    public OpenSergoClient getOrCreateClient(String host, int port, OpenSergoConfig config) {
```
Comment (sczyh30) [inline src/main/java/io/opensergo/OpenSergoClientManager.java:79]: Pay attention to thread-safety here.

File: src/main/java/io/opensergo/OpenSergoClient.java [hunk 2]
```diff
@@ -48,12 +48,39 @@ public class OpenSergoClient implements AutoCloseable {
...
+        private int port;
+        private OpenSergoConfig openSergoConfig;
+
+        public OpenSergoClient.Builder endpoint(String host, int port) {
+            this.host = host;
+            this.port = port;
+            return this;
+        }
+
+        public OpenSergoClient.Builder openSergoConfig(OpenSergoConfig openSergoConfig) {
```
Comment (sczyh30) [inline src/main/java/io/opensergo/OpenSergoClient.java:63]: 后面我们需要优化 builder 里面的元素，是不是可以把 OpenSergoConfig 里面的配置平铺到 builder 里面处理，让用户更易用？

**Review by sczyh30 — APPROVED** [review]
LGTM. I'll improve the details later.

**Comment by sczyh30** [comment]
Thanks!

File: src/main/java/io/opensergo/OpenSergoClient.java (hunk 2 above)
Comment (jnan806) [reply src/main/java/io/opensergo/OpenSergoClient.java:63]: 是把 OpenSergoConfig 取消掉吗？

**Pull request merged**